### ISSUE TITLE
tf.keras.metrics.categorical_crossentropy uses wrong class count for label smoothing with axis zero

### ISSUE THREAD

**Issue by ALinrunrun** [tensorflow/tensorflow#127249]
### Issue type

Bug

### Have you reproduced the bug with TensorFlow Nightly?

Yes

### Source

source

### TensorFlow version

tf 2.21.0, also reproduced on tf 2.22.0-dev20260904

### Custom code

Yes

### OS platform and distribution

Linux Ubuntu 22.04

### Python version

Python 3.13.5

### Current behavior?

`tf.keras.metrics.categorical_crossentropy` returns an incorrect forward value when `label_smoothing` is used with `axis=0`.

In the reproducer below, the class axis is `axis=0`, whose size is 3. Each target column sums to one, so label smoothing with `label_smoothing=0.25` should distribute the smoothing mass as `0.25 / 3` across the three classes.

However, the returned loss appears to use the final dimension size 2 as the class count instead. This changes the smoothed labels and therefore changes the forward loss.

The expected weighted loss is approximately `250.08333333333363`, but TensorFlow returns `268.25000000000034`.

Expected behavior?

When `axis=0`, `tf.keras.metrics.categorical_crossentropy` should use the size of axis 0 as the class count for label smoothing. For the reproducer below, the weighted loss should be close to `250.08333333333363`, not `268.25000000000034`.

### Standalone code to reproduce the issue

```shell
import os
os.environ["CUDA_VISIBLE_DEVICES"] = ""
os.environ["OMP_NUM_THREADS"] = "1"
os.environ["TF_NUM_INTRAOP_THREADS"] = "1"
os.environ["TF_NUM_INTEROP_THREADS"] = "1"

import tensorflow as tf

def target(t):
    y_true = tf.constant(
        [[1, 0],
         [0, 0],
         [0, 1]],
        dtype=t.dtype,
    )

    y_pred = tf.reshape(
        tf.stack([
            t,
            -t,
            2 * t + 1,
            t / 2,
            -t + 2,
            3 * t - 1,
        ]),
        (3, 2),
    )

    loss = tf.keras.metrics.categorical_crossentropy(
        y_true,
        y_pred,
        from_logits=True,
        label_smoothing=0.25,
        axis=0,
    )

    return tf.reduce_sum(tf.constant([1, 3], dtype=t.dtype) * loss)

x = tf.constant(-20.0, dtype=tf.float64)

actual = target(x)
expected = 250.08333333333363

print("actual:", actual.numpy())
print("expected:", expected)

if abs(float(actual) - expected) > 1e-9:
    print("BUG REPRODUCED: tf.keras.metrics.categorical_crossentropy uses the wrong class count for label smoothing with axis zero")
else:
    print("not reproduced")
```

### Relevant log output

```shell
actual: 268.25000000000034
expected: 250.08333333333363
BUG REPRODUCED: tf.keras.metrics.categorical_crossentropy uses the wrong class count for label smoothing with axis zero
```

**Comment by Kayyuri** [comment]
Hi @ALinrunrun,  Thank you for raising your concern

 I tested the provided code with the latest tf-nightly(`2.22.0-dev20260910`), and the issue appears to be fixed. `categorical_crossentropy` now correctly uses the size of `axis=0` as the class count when applying label smoothing. I’m no longer able to reproduce the incorrect loss value with the latest TensorFlow build. Could you please check whether this issue is still reproducible with the latest `tf-nightly` version and let us know if you are still experiencing the same issue?
I am attaching the [gist](https://colab.sandbox.google.com/gist/Kayyuri/f6147790c2dfc755fde051170508ea91/127249.ipynb) file for your reference.

Thanks!
